File: sdks/tokenizer-sdk/src/accounts/buyoutOffer.ts
/**
 * BuyoutOffer account deserializer.
 *
 * #[repr(C)] layout (288 bytes):
 *   0: account_key (u8)
 *   1: version (u8)
 *   2: buyer ([u8;32])
 *  34: asset ([u8;32])
 *  66: padding (6 bytes)
 *  72: price_per_share (u64)
 *  80: accepted_mint ([u8;32])
 * 112: escrow ([u8;32])
 * 144: treasury_disposition (u8)
 * 145: terms_hash ([u8;32])
 * 177: broker ([u8;32])
 * 209: padding (1 byte)
 * 210: broker_bps (u16)
 * 212: padding (4 bytes)
 * 216: broker_amount (u64)
 * 224: minted_shares (u64)
 * 232: shares_settled (u64)
 * 240: treasury_amount (u64)
 * 248: status (u8)
 * 249: is_council_buyout (u8)
 * 250: padding (6 bytes)
 * 256: expires_at (i64)
 * 264: created_at (i64)
 * 272: updated_at (i64)
 * 280: bump (u8)
 * 281: rent_payer ([u8;32])
 */
import { type Address, getStructDecoder, transformDecoder } from "gill";
import { AccountKey } from "../constants.js";
import { addr, bytes32, i64d, pad, u8d, u16d, u64d } from "./decode.js";

export const BUYOUT_OFFER_SIZE = 320;

export interface BuyoutOffer {
  accountKey: number;
  version: number;
  buyer: Address;
  asset: Address;
  pricePerShare: bigint;
  acceptedMint: Address;
  escrow: Address;
  treasuryDisposition: number;
  termsHash: Uint8Array;
  broker: Address;
  brokerBps: number;
  brokerAmount: bigint;
  mintedShares: bigint;
  sharesSettled: bigint;
  treasuryAmount: bigint;
  status: number;
  isCouncilBuyout: boolean;
  expiresAt: bigint;
  createdAt: bigint;
  updatedAt: bigint;
  bump: number;
  rentPayer: Address;
}

const rawDecoder = getStructDecoder([
  ["accountKey", u8d],
  ["version", u8d],
  ["buyer", addr],
  ["asset", addr],
  ["_p0", pad(6)],
  ["pricePerShare", u64d],
  ["acceptedMint", addr],
  ["escrow", addr],
  ["treasuryDisposition", u8d],
  ["termsHash", bytes32],
  ["broker", addr],
  ["_p1", pad(1)],
  ["brokerBps", u16d],
  ["_p2", pad(4)],
  ["brokerAmount", u64d],
  ["mintedShares", u64d],
  ["sharesSettled", u64d],
  ["treasuryAmount", u64d],
  ["status", u8d],
  ["isCouncilBuyout", u8d],
  ["_p3", pad(6)],
  ["expiresAt", i64d],
  ["createdAt", i64d],
  ["updatedAt", i64d],
  ["bump", u8d],
  ["rentPayer", addr],
  ["_p4", pad(7)],
]);

export const buyoutOfferDecoder = transformDecoder(
  rawDecoder,
  ({ _p0, _p1, _p2, _p3, _p4, isCouncilBuyout, termsHash, ...rest }) => ({
    ...rest,
    isCouncilBuyout: isCouncilBuyout !== 0,
    termsHash: termsHash as Uint8Array,
  }),
);

export function decodeBuyoutOffer(data: Uint8Array): BuyoutOffer {
  if (data.length < BUYOUT_OFFER_SIZE) {
    throw new Error(`BuyoutOffer: expected ${BUYOUT_OFFER_SIZE} bytes, got ${data.length}`);
  }
  const accountKey = data[0];
  if (accountKey !== AccountKey.BuyoutOffer) {
    throw new Error(`BuyoutOffer: invalid account key ${accountKey}`);
  }
  return buyoutOfferDecoder.decode(data);
}

File: programs/tokenizer/src/state/buyout_offer.rs
use pinocchio::Address;

use super::BUYOUT_OFFER_SEED;

/// Full buyout offer for an asset.
/// PDA: ["buyout_offer", asset.key(), buyer.key()]
#[repr(C)]
pub struct BuyoutOffer {
    pub account_key: u8,              // 0
    pub version: u8,                  // 1
    pub buyer: [u8; 32],              // 2
    pub asset: [u8; 32],              // 34
    // pad 6 bytes                    // 66-71
    pub price_per_share: u64,         // 72
    pub accepted_mint: [u8; 32],      // 80
    pub escrow: [u8; 32],             // 112
    pub treasury_disposition: u8,     // 144
    pub terms_hash: [u8; 32],         // 145
    pub broker: [u8; 32],             // 177
    // pad 1 byte                     // 209
    pub broker_bps: u16,              // 210
    // pad 4 bytes                    // 212-215
    pub broker_amount: u64,           // 216
    pub minted_shares: u64,           // 224
    pub shares_settled: u64,          // 232
    pub treasury_amount: u64,         // 240
    pub status: u8,                   // 248
    pub is_council_buyout: u8,        // 249
    // pad 6 bytes                    // 250-255
    pub expires_at: i64,              // 256
    pub created_at: i64,              // 264
    pub updated_at: i64,              // 272
    pub bump: u8,                     // 280
    pub rent_payer: [u8; 32],         // 281
    // trailing pad updated by repr(C)
}

impl BuyoutOffer {
    pub const LEN: usize = core::mem::size_of::<Self>();

    #[inline(always)]
    pub unsafe fn load(data: &[u8]) -> &Self {
        &*(data.as_ptr() as *const Self)
    }

    #[inline(always)]
    pub unsafe fn load_mut(data: &mut [u8]) -> &mut Self {
        &mut *(data.as_mut_ptr() as *mut Self)
    }

    pub fn derive_pda(
        asset: &[u8; 32],
        buyer: &[u8; 32],
        program_id: &Address,
    ) -> (Address, u8) {
        Address::find_program_address(
            &[BUYOUT_OFFER_SEED, asset, buyer],
            program_id,
        )
    }
}
